Mage/src/main/java/mage/game/combat/CombatGroup.java

```diff
@@ -322,42 +274,47 @@ private void multiBlockerDamage(Player player, boolean first, Game game) {
                 }
             }
             Map<UUID, Integer> assigned = new HashMap<>();
+            List<MultiAmountMessage> damageDivision = new ArrayList<>();
+            List<UUID> blockersCopy = new ArrayList<>(blockers);
             if (blocked) {
-                boolean excessDamageToDefender = true;
-                for (UUID blockerId : new ArrayList<>(blockerOrder)) { // prevent ConcurrentModificationException
+                int remainingDamage = damage;
+                for (UUID blockerId : blockers) {
                     Permanent blocker = game.getPermanent(blockerId);
                     if (blocker != null) {
-                        int lethalDamage = getLethalDamage(blocker, attacker, game);
-                        if (lethalDamage >= damage) {
-                            if (!oldRuleDamage) {
-                                assigned.put(blockerId, damage);
-                                damage = 0;
-                                break;
-                            } else if (damage == 0) {
-                                break;
-                            }
-                        }
-                        int damageAssigned = 0;
-                        if (!oldRuleDamage) {
-                            damageAssigned = player.getAmount(lethalDamage, damage, "Assign damage to " + blocker.getName(), game);
-                        } else {
-                            damageAssigned = player.getAmount(0, damage, "Assign damage to " + blocker.getName(), game);
-                            if (damageAssigned < lethalDamage) {
-                                excessDamageToDefender = false; // all blockers need to have lethal damage assigned before it can trample over to the defender
-                            }
-                        }
-                        assigned.put(blockerId, damageAssigned);
-                        damage -= damageAssigned;
+                        int defaultDamage = Math.min(remainingDamage, blocker.getLethalDamage(attacker.getId(), game));
+                        remainingDamage -= defaultDamage;
+                        String message = String.format("%s, P/T: %d/%d",
+                                blocker.getLogName(),
+                                blocker.getPower().getValue(),
+                                blocker.getToughness().getValue());
+                        damageDivision.add(new MultiAmountMessage(message, 0, damage, defaultDamage));
+                    }
+                }
+                List<Integer> amounts;
+                if (hasTrample(attacker)){
+                    amounts = player.getMultiAmountWithIndividualConstraints(Outcome.Damage, damageDivision, damage-remainingDamage, damage, MultiAmountType.DAMAGE_TRAMPLE, game);
+                    int trampleDamage = damage-(amounts.stream().mapToInt(x -> x).sum());
```

Comment:
There are possible infinite or big values like Integer.MAX_VALUE. If it’s a direct values from permanents/players then I recommend to use overflow protection methods like overflowInc and overflowDec from CardUtil and min/max protection to make sure it’s positive only values.

Reply:
Big values should be fine, I don't think there's any possible overflow issues.

The only direct value is the original damage amount. The code for `defaultDamage` uses `min` to ensure `remainingDamage` never go below zero. Unless the MultiAmount constraints are somehow broken, the sum of the amounts will always be less than or equal to the damage to be dealt and as such no overflow protection should be necessary. I can add it in anyway just in case if you'd like?

However, it might be wrong if the attacker has negative power. I'll check that.